### PR TITLE
Modify RazorPreCompileModule to use an instance of memory cache specific to the application's target framework and configuration.

### PR DESCRIPTION
**Pull request by pranavkm**
cc @davidfowl  \ @anurse

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+                    _memoryCacheLookup[cacheKey] = memoryCache;
+                }
+            }
+
+            var viewCompiler = new RazorPreCompiler(
+                context,
+                _loadContext,
+                fileProvider,
+                memoryCache)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs]: MemoryCache is limited to using string keys - so we're stuck with either creating a composite key using the tfm and configuration - or varying the instance in itself. This seemed cleaner.

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            public bool Equals(PrecompilationCacheKey other)
+            {
+                return
+                    other.TargetFramework == TargetFramework &&
+                    string.Equals(other.Configuration, Configuration, StringComparison.Ordinal);
+            }
+
+            public override int GetHashCode()
+            {
+                return 13 * Configuration.GetHashCode() + TargetFramework.GetHashCode();
```
Comment (anurse) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs]: HashCodeCombiner? Now in the handy-dandy `Microsoft.Framework.HashCodeCombiner.Sources` package :)

**Comment by anurse** [comment]
:shipit:

The cache-per-(config,tfm) thing is a little odd but I think you're right that it's easier than changing the whole cache interface.

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        private readonly Dictionary<PrecompilationCacheKey, MemoryCache> _memoryCacheLookup =
+            new Dictionary<PrecompilationCacheKey, MemoryCache>();
+
+        /// <summary>
+        /// Instantiates a new <see cref="RazorPreCompileModule"/> instance.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceProvider"/> for the application.</param>
+        public RazorPreCompileModule(IServiceProvider services)
+        {
+            _loadContext = services.GetRequiredService<IAssemblyLoadContext>();
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs]: Do you need this?

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs]: That should go away once this PR goes in - https://github.com/aspnet/Mvc/pull/2961

**Pull request closed without merging**